### PR TITLE
fix(core): reset existing properties in Manifest#load

### PR DESCRIPTION
**Pull request by bgotink**
**What's the problem this PR addresses?**

As mentioned by @merceyz  in https://github.com/yarnpkg/berry/pull/2245#issuecomment-748990256, #1775 introduced another bug because `Manifest#load` unconditionally adds to the `workspaceDefinitions` array which leads to duplicate entries in workspaces (and only in workspaces AFAICT) when constraints are fixed.

**How did you fix it?**

Reset existing manifest properties during `Manifest#load`.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-core/sources/Manifest.ts [hunk 1]
```diff
@@ -356,6 +356,9 @@ export class Manifest {
         continue;
       }
 
+      if (this.workspaceDefinitions.some(({pattern}) => pattern === entry))
+        continue;
+
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/Manifest.ts:361]: Could we use a Set with the patterns in it to make the lookup faster?

Comment (bgotink) [reply packages/yarnpkg-core/sources/Manifest.ts]: Makes sense.

I've added the `Set` to the `load` function to keep the API of `Manifest` as is.

**Comment by merceyz** [comment]
> Note that this does change the behaviour of yarn in that it will automatically remove duplicate workspaces entries whenever you run yarn install.

Perhaps we should we do that, we normalize other fields so might as well take this one as well

File: packages/yarnpkg-core/sources/Manifest.ts [hunk 2]
```diff
@@ -350,12 +350,17 @@ export class Manifest {
         ? data.workspaces.packages
         : [];
 
+    const existingWorkspaces = new Set(this.workspaceDefinitions.map(({pattern}) => pattern));
```
Comment (arcanis) [inline packages/yarnpkg-core/sources/Manifest.ts:353]: Couldn't we just reset `this.workspaceDefinitions` to an empty array? This way it wouldn't change its type / content.

Comment (bgotink) [reply packages/yarnpkg-core/sources/Manifest.ts:353]: That's also possible.

There are some other problems with the `Manifest#load` function, e.g. it resets some properties but it doesn't reset `dependencies`/`optionalDependencies`/`peerDependencies`/`devDependencies`/`dependenciesMeta`/`peerDependenciesMeta` and `resolutions`. This means it's possible to add/update values in these objects, but impossible to remove them by reloading raw JSON data. The load method also doesn't support removing `preferUnplugged`.

All of those properties should be reset, otherwise we'll just end up getting more bugs like "constraints can't remove resolutions/dependenciesMeta/peerDependenciesMeta/preferUnplugged... via `gen_enforced_field`"

In short, this method wasn't written to be used on an already loaded manifest, yielding all kind of shenanigans in constraints because we do use it on an already loaded manifest.

**Comment by arcanis** [comment]
Hooray for more consistency! ✨
